src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets

```diff
@@ -177,37 +180,27 @@
-       copied via RunTests script so VS can work when the test dir is initially clean.
-       -->
+  <!-- This copies the same files as the RunTests script, when running locally.
+       This achieves two things: 
+	   1) Allows VS to work when the test dir is initially clean, since VS does not invoke the runner script
+        because it needs to attach to the XUnit process when debugging.
+	   2) For incremental build of any kind, ensures that any test dependency which is updated gets copied
+       (thanks to smarter copy logic) -->
   <PropertyGroup Condition="'$(BuildingInsideVisualStudio)'=='true'">
-    <PrepareForRunDependsOn>$(PrepareForRunDependsOn);CopyDefaultTestAssetsForVS</PrepareForRunDependsOn>
+    <PrepareForRunDependsOn>$(PrepareForRunDependsOn);CopyDependenciesToTestDirectory</PrepareForRunDependsOn>
   </PropertyGroup>
-  <Target Name="CopyDefaultTestAssetsForVS" DependsOnTargets="CopyTestToTestDirectory;CopySupplementalTestData">
-     <!-- This was copied from RunTestsForProject in tests.targets
-          The RunTestsForProject target does not execute in VS context and would be confused by the script based runner.
-          _TestCopyLocalByFileNameWithoutDuplicates are the precise items that are fed to the runner script generation code.
-     -->
-    <ItemGroup>
-      <!-- Not all platforms can use the .ni.dlls that come from packages.  If TestWithoutNativeImages is specified, we'll exclude them from copy generation.
-           If we end up needing this for any other sorts of filtering, we'll want to add a list of filtered extensions to be matched on EndsWith.  -->
-      <_IncludedFileForTestsInVS Include="@(_TestCopyLocalByFileNameWithoutDuplicates)"
-                                 Condition="'$(TestWithoutNativeImages)' != 'true' Or !$([System.String]::Copy('%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)').EndsWith('.ni.dll'))" >
-        <PackageRelativePath Condition="'%(_TestCopyLocalByFileNameWithoutDuplicates.NugetPackageId)' != ''">$([System.String]::Copy('%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)').Replace('$(PackagesDir)',''))</PackageRelativePath>
-        <UseAbsolutePath Condition="'$(TestWithLocalLibraries)'=='true'">$([System.String]::Copy('%(_TestCopyLocalByFileNameWithoutDuplicates.SourcePath)').StartsWith('$(BinDir)'))</UseAbsolutePath>
-        <DestinationPath>$(TestPath)\%(Filename)%(Extension)</DestinationPath>
-      </_IncludedFileForTestsInVS>
-      <_IncludedFileForTestsInVs Remove="@(_IncludedFileForTestsInVS)" Condition="Exists('%(DestinationPath)')" />
-    </ItemGroup>
-
+  <Target Name="CopyDependenciesToTestDirectory"
+          Condition="'$(DisableCopyTestToTestDirectory)' != 'true' And '$(ArchiveTests)' != 'true'"
+          DependsOnTargets="CopyTestToTestDirectory;DiscoverTestDependencies">
     <Copy
-      SourceFiles="@(_IncludedFileForTestsInVS  -> '%(SourcePath)')"
```

Comment:
The main benefit of copying the files at execution time for Helix is that it reduces the build time by reducing the sizes of the archives uploaded to Azure.  I just tested archiving the entire test runtime directory for System.Collections.Tests, and the archive was 19.8 MB, compared to 1.30 MB with the current implementation. There are about 200 projects, so in total that would be a difference of roughly 4 GB vs. 260 MB.  It also took several seconds longer to create the archive just for the one project, so it would add build time there too.

The reason that we're copying files at build time when building locally is to ensure correct functionality for incremental build scenarios where the test runtime directory already exists.  This behavior is what @weshaggard and @MattGal proposed in #1163.

The compromise to ensure the same functionality in Helix vs. non-Helix which is implemented here is to use the same item group for generating the list of files to copy at build time locally and to put in the test execution scripts for copying at execution time in Helix.